docs/articles/semantic_chunking.md

```diff
@@ -0,0 +1,537 @@
+        sentences = self.extractor.run(text)
+
+        if isinstance(sentences, list):
+            return sentences
+
+        # Fallback: extract sentences heuristically
+        return [s.strip() for s in text.split('.') if s.strip()]
+
+    def split_text(self, text: str) -> List[str]:
+        """Extract propositions and chunk them accordingly."""
+        propositions = self.get_propositions(text)
+        return self._chunk_propositions(propositions)
+
+    def split_documents(self, documents: List[Document]) -> List[Document]:
+        """Split documents into chunks."""
+        split_docs = []
+        for doc in documents:
+            chunks = self.split_text(doc.page_content)
+            for i, chunk in enumerate(chunks):
+                metadata = doc.metadata.copy()
+                metadata.update({"chunk_index": i})
+                split_docs.append(Document(page_content=chunk, metadata=metadata))
+        return split_docs
+
+    def _chunk_propositions(self, propositions: List[str]) -> List[str]:
+        chunks = []
+        current_chunk = []
+        current_size = 0
+
+        for prop in propositions:
+            prop_size = len(prop)
+            if current_size + prop_size > self.chunk_size and current_chunk:
+                chunks.append(" ".join(current_chunk))
+                current_chunk = []
+                current_size = 0
+            current_chunk.append(prop)
+            current_size += prop_size
+
+        if current_chunk:
+            chunks.append(" ".join(current_chunk))
+
+        return chunks
+```
+
+## Evaluation
+
+We conducted experiments with the above three variations of semantic chunking on two datasets and embedding models.
+...to see how they performed in terms of...
+
+***Datasets***
+
+1. [HotpotQA](https://huggingface.co/datasets/hotpotqa/hotpot_qa?row=16)
+2. [SQUAD](https://huggingface.co/datasets/squad?row=0)
+
+***Embedding models***
+
+1. [BAAI/bge-small-en-v1.5](https://huggingface.co/BAAI/bge-small-en-v1.5)
+2. BAAI/bge-small-en-v1.5 with reranker
+
+Here are the results.
+
+![Evaluation results](../assets/use_cases/semantic_chunking/table.png)
+
+![Results summary](../assets/use_cases/semantic_chunking/graph5.png)
+
+From the above results, we can summarize.
+
+- LLM based semantic chunking performs best overall, especially in faithfulness and handling complex queries as LLMs can understand context and semantics at a deep level, allowing for more intelligent chunking. But it has high latency numbers as LLM inference takes time.
+- Embedding similarity based semantic chunking shows strong performance in relevancy and a good balance of precision and recall.
+- Clustering based semantic chunking is more variable, potentially excelling in specific scenarios but possibly struggling with others. It has slightly higher latency numbers compared to embedding based method.
+- The reranker improves performance across all methods.
+
+## Wrapping up
+
+
+## References

```

Comment:
@Ashish-Abraham
if you want to find a place in the text where these references are directly relevant, we can include them. Otherwise we should drop them.

Reply:
We can drop them then. I don't see where they are directly relevant  in my content. They are just resources I read before I started working on my article.